### PR TITLE
Dummy PR for DDD diffs

### PR DESCRIPTION
**Pull request by hlandau**
This is to allow people to leave comments on the diffs.

### CODE REVIEW

File: ddd/ddd-03-fd-blocking.DISCUSSION.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      compatible way to pass the destination endpoint to libssl. Once the FD
+      is passed to libssl, libssl can obtain the destination endpoint if
+      it needs to know by calling getpeername(3).
+
+    - As for ddd-01, it is assumed that the application spends a lot of time
+      in calls to libssl (as it is a blocking application) or at least calls it
+      regularly. Since synchronous users of libssl are probably fairly simple
+      this may be true quite often. If it is not the case, either a
+      BIO_pump/SSL_pump call must be made regularly via some means, or
+      thread-assisted operation should be used.
```
Comment (t8m) [inline ddd/ddd-03-fd-blocking.DISCUSSION.md:25]: It's more problematic than this. There is another problem of being blocked on read/readmsg() call on a blocking socket - the workaround could be to always temporarily set the socket to nonblocking mode internally before we call read/readmsg() on the socket and use poll() on the socket with a timeout. Probably having an internal thread could be another workaround.

Comment (hlandau) [reply ddd/ddd-03-fd-blocking.DISCUSSION.md:25]: There are a few options here. In thread-assisted mode we have a lot of options. For the non-thread-assisted case:

- We could always poll before calling recvmsg to guarantee it doesn't block on the next call.
- We could use recvmsg with MSG_DONTWAIT where it is supported.
- We could set the FD as nonblocking. This could be a fallback for platforms which can't support MSG_DONTWAIT. This may have application-visible side effects if the application passed the FD, but it seems like it would be extremely unusual for an application to give a FD to libssl but also try and do things on it which are affected by O_NONBLOCK.

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      we ever end up needing multiple sockets, we can provide an advanced
+      interface which applications can choose to implement if they wish. In the
+      worst case they can forego some functionality if they don't implement it.
+      But to be clear, there is no immediate identified need for multiple
+      sockets at this time.
+
+  - The application needs to know when to listen for each of
+    POLLIN/POLLOUT/POLLERR on the file descriptor it has been told about.
+
+    CHANGE: The function `BIO_get_fd` is changed to `BIO_get_poll_fd` as the
```
Comment (mattcaswell) [inline ddd/ddd-02-conn-nonblocking.DISCUSSION.md:52]: Not sure I understand the reasoning behind this. In what way is the fd returned by `BIO_get_poll_fd` different from the fd returned by `BIO_get_fd`?

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    - The QUIC stack may need to randomly send frames (e.g. keepalives)
+      even when the application is not calling SSL_read/SSL_write for a long
+      time (e.g. an hour).
+
+    - So arrangements must be made for some call to libssl to be made
+      periodically.
+
+      PROPOSED SOLUTION: A call, BIO_pump (/SSL_pump) is introduced which can be used
+      when an application wishes to allow libssl to process events without
+      calling read/write functions.
```
Comment (mattcaswell) [inline ddd/ddd-02-conn-nonblocking.DISCUSSION.md:110]: In DTLS we have `DTLSv1_handle_timeout()` which does this. In openssl/openssl#17184 I introduced `SSL_tick()` to do this.

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    - It is possible this can be handled automatically on platforms where
+      threads can be used but this cannot be assumed on all platforms.
+
+    - Worst case, provide a "get wakeup time" function along with the "get FD"
+      function which indicates the deadline by which libssl should next be
+      called.
+
+      PROPOSED SOLUTION: A function BIO_get_timeout is introduced which returns
+      when libssl next wants to be called.
```
Comment (mattcaswell) [inline ddd/ddd-02-conn-nonblocking.DISCUSSION.md:122]: `BIO_get_timeout` doesn't seem like a good fit. I would expect this to be an "SSL" function not a "BIO" function. The DTLS code already has `DTLSv1_get_timeout()` in openssl/openssl#17184 I proposed `SSL_get_next_tick()`.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+pure state machine which does no I/O itself.
+
+The problems with adapting this to QUIC include:
+
+- QUIC uses UDP rather than TCP and the contents of its UDP packets are
+  not concatenatable, whereas TCP uses a byte stream model. So existing
+  applications based around this approach will assume they can pull data out
+  of the memory BIOs in arbitrary-length chunks and feed them to the
+  network, which is not the case. Moreover, if libssl just writes the UDP
+  datagram payload to the memory BIO for each UDP datagram it wants ton
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:23]: nit: typo (ton)

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    application.
+
+  - Looking at the big picture, the *reason* an application wants to use
+    a memory BIO on the network side is so it can use libssl as a pure
+    state machine and be in control of all I/O.
+
+    Since supporting the exact same interface (memory BIOs with bytestream
+    semantics) is infeasible for the reasons given above, the next best thing we
+    can do is to ensure that we accommodate this use case for applications, so
+    that there is some rougly analagous solution available to them. This may not
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:44]: nit: typo (rougly)

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    BIO_read calls with the same lengths returned).
+
+  - There are some caveats for this idea. In order for this to work:
+
+    - applications must ensure there is a guaranteed 1:1 correspondance
+      between send/recv and BIO_read/BIO_write calls; and
+
+    - the buffer passed to BIO_read() and recv() must always be big
+      enough to receive the largest possible packet, else silent
+      truncation occurs.
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:68]: We can't do much to help applications in calling`recv` this is outside of OpenSSL. For a `BIO_dgram_mem` we can fail if the buffer passed to `BIO_read()` isn't big enough. We can also supply a new BIO ctrl to find out the size of the next packet in the mem buffer.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    include:
+
+    - Using a maximum MTU of 1472 by default; then any application using
+      old BIO_read/BIO_write calls will work so long as it uses a buffer
+      of at least 1472 bytes (and there is a 1:1 correspondance of BIO calls and
+      syscalls).
+
+    - Applications which can support larger MTUs (e.g. a network with
+      jumbo frames) could indicate support via some method TBD, such as
+      capability flags (see below).
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:97]: Can't `BIO_dgram_mem` just support `BIO_CTRL_DGRAM_SET_MTU` in the same way that `BIO_dgram` does. The application can then set it to whatever value is appropriate (or otherwise let it default to 1472). `BIO_dgram_mem` should support the same MTU querying ctrls that `BIO_dgram` supports, e.g. `BIO_CTRL_DGRAM_QUERY_MTU`

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+    In this model, for the purposes of compatibility and minimising changes:
+
+    - BIO_ctrl_get_write_guarantee (BIO_C_GET_WRITE_GUARANTEE) would always
+      return either 0 or a value >=1472. (This is a change from BIO_dgram,
+      which does not implement BIO_C_GET_WRITE_GUARANTEE.)
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:103]: Not sure I understand the need for this?

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+      /*
+       * A flag settable on BIO_dgram_mem. A BIO_read call will not truncate,
+       * but instead fail if the frame is too large for the provided buffer. The
+       * application must call BIO_ctrl_pending() to determine the necessary
+       * buffer size and try again.
+       */
+      int BIO_dgram_set_no_trunc(bio, 1);
+
+      /*
+       * Returns libssl's idea of what our maximum UDP payload size is.
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:125]: This is odd....asking a BIO about what libssl's idea of something is. Libssl should surely ask the BIO what the MTU is and use that not the other way around.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    larger changes):
+
+    ```c
+      /*
+       * A flag settable on BIO_dgram_mem. A BIO_read call will not truncate,
+       * but instead fail if the frame is too large for the provided buffer. The
+       * application must call BIO_ctrl_pending() to determine the necessary
+       * buffer size and try again.
+       */
+      int BIO_dgram_set_no_trunc(bio, 1);
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:122]: Not sure why we need this. Why not just make this the default (and only) behaviour? I'm struggling to see why you would ever want it to truncate.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+        BIO_get_dgram_origin (BIO_CTRL_DGRAM_GET_PEER)
+        BIO_set_dgram_dest   (BIO_CTRL_DGRAM_SET_ADDR)
+        BIO_set_dgram_origin (BIO_CTRL_DGRAM_SET_PEER)
+      ```
+
+      The get functions can be called after a BIO_read and yield information
+      on the last datagram received. Likewise, the set functions can be called
+      before a BIO_write and control the source and destination of the write.
+
+      This API should be fine.
```
Comment (mattcaswell) [inline ddd/ddd-05-mem-nonblocking.DISCUSSION.md:148]: openssl/openssl#17184 introduces explicit params for the peer and local BIO_ADDR values in the proposed BIO_sendmmsg and BIO_recvmmsg functions.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
We should bring these proposals and openssl/openssl#17184 together. That PR suggests some things which have an overlap with what is proposed here. It also proposes new APIs that will be needed for the future (not necessarily for MVP). We should figure out how your suggestions and that PR fit together.

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md (hunk 3 above)
Comment (hlandau) [reply ddd/ddd-02-conn-nonblocking.DISCUSSION.md:110]: Good point. I'll change this to `BIO_tick`/`SSL_tick`. I think our ideas are the same. In future maybe DTLSv1_handle_timeout  can be deprecated in favour of tick.

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md (hunk 2 above)
Comment (hlandau) [reply ddd/ddd-02-conn-nonblocking.DISCUSSION.md:52]: The idea is that we may not want to guarantee anything about the FD that should be polled (whether it is a socket, etc.) in future. If the API contract is just "this is a pollable FD, trying to do anything else with it is undefined", this keeps our  options open.

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md (hunk 4 above)
Comment (hlandau) [reply ddd/ddd-02-conn-nonblocking.DISCUSSION.md:122]: I'd be totally happy with either of these.

Where a `BIO_*` function is mentioned in this proposal generally the implication is it just calls the equivalent `SSL_*` function for `BIO_f_ssl` and is otherwise unimplemented, it's just a helper method. But we can also make the client do it itself and call the `SSL_*` function directly, it doesn't really matter.

Maybe we can deprecate `DTLSv1_get_timeout` and replace it with `SSL_get_next_tick` in future.

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md (hunk 2 above)
Comment (hlandau) [reply ddd/ddd-02-conn-nonblocking.DISCUSSION.md:52]: Of course, there's a valid question as to whether this is a useful option to keep open. Are we likely, in future, to shift to some different QUIC design such that it's useful to return some unspecified non-socket FD here? (Think of the things the Linux kernel offers, eventfds, timerfds, etc.) Possibly not, though if we discover new requirements after MVP (for advanced features like connection migration etc.), it seems like a conceivable possibility. API design instincts tell me we should keep this as vague as possible if the client doesn't need a specific guarantee.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 7 above)
Comment (hlandau) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:68]: >We can also supply a new BIO ctrl to find out the size of the next packet in the mem buffer.

Definitely, if we don't even already have one.

>For a BIO_dgram_mem we can fail if the buffer passed to BIO_read() isn't big enough.

I like this idea, but this causes the semantics of a `BIO_dgram` (which is just a wrapper around `recvfrom`) to differ from those of `BIO_dgram_mem`, so it doesn't really seem like an option. The entire point is for these to have the same semantics.

>We can't do much to help applications in calling `recv` this is outside of OpenSSL.

There's certainly a valid discussion to be had about just *how* far we go to minimise changes necessary. Ordinarily we would expect application changes to be needed here due to the change from datagram semantics to bytestream semantics. If certain conditions are met, as discussed in the document, applications *might* be able to get away without any changes. If those conditions aren't met, the application will certainly have to adapt, as there is nothing we can possibly do to make it work without further changes to their code.

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md (hunk 2 above)
Comment (mattcaswell) [reply ddd/ddd-02-conn-nonblocking.DISCUSSION.md:52]: Ok, this makes sense. It's quite similar to the ASYNC "wait_fd" mechanism that we have:

https://www.openssl.org/docs/manmaster/man3/ASYNC_WAIT_CTX_set_wait_fd.html

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 9 above)
Comment (hlandau) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:103]: My motivation for this is that if you return a value `x` below 1472, it will cause problems with application code that does something like `BIO_write(buf, min(..., BIO_ctrl_get_write_guarantee()))`.

Essentially the idea here is that the BIO_dgram_mem can either accept another dgram or it can't. It never "partially" accepts dgrams written to it, so it makes no sense for it to claim it can accept dgrams but only if they're smaller than the MTU.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 11 above)
Comment (hlandau) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:122]: See above, we need to do this to match the semantics of the existing `BIO_dgram` which just wraps `recvfrom`, which truncates. This behaviour of the Berkeley sockets API has always really irked me, but we're stuck with it.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 10 above)
Comment (hlandau) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:125]: For `BIO_dgram` we have `BIO_CTRL_DGRAM_GET/SET_MTU` which `libssl` can make use of to find an MTU communicated by the application. So I guess the application would create the `BIO_dgram_mem`, which defaults to an MTU (more accurately, maximum UDP payload length) of 1472, and can then set a larger MTU using `BIO_CTRL_DGRAM_SET_MTU` if it wishes.

IIRC, I believe where I was going with `BIO_get_max_dgram_len` is that we should have a function that more capable applications can use to ask the QUIC stack what maximum UDP payload length it would *like* to be able to use. A better name for this would probably be `SSL_get_preferred_max_dgram_len`.

So basically, for a basic client:
- `char buf[1472 (or larger)];`
- never calls `BIO_CTRL_DGRAM_SET_MTU`, `BIO_dgram_mem` defaults to 1472
- libssl calls `BIO_CTRL_DGRAM_GET_MTU` and determines it can use a UDP payload size of 1472

For a better client:
```
ssize_t maxlen = SSL_get_preferred_max_dgram_len(ssl);
...
char *buf = malloc(maxlen);
BIO_CTRL_DGRAM_SET_MTU(maxlen);
```

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md (hunk 4 above)
Comment (mattcaswell) [reply ddd/ddd-02-conn-nonblocking.DISCUSSION.md:122]: In that case I'd suggest we list the `SSL_*` function first as the primary API for this. 

DTLSv1_get_timeout is a bad name anyway (it works for both DTLSv1 and DTLSv1.2). This is true for a number of the DTLS API functions. Whether it warrants deprecation, I'm not sure.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 7 above)
Comment (mattcaswell) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:68]: > I like this idea, but this causes the semantics of a BIO_dgram (which is just a wrapper around recvfrom) to differ from those of BIO_dgram_mem, so it doesn't really seem like an option. The entire point is for these to have the same semantics.

I wonder how much this matters? It means that BIO_dgram_mem is slightly stricter that BIO_dgram - but it is stricter in the case where BIO_dgram will silently fail which is probably unwanted behaviour.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 9 above)
Comment (mattcaswell) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:103]: BIO_dgram has `BIO_CTRL_DGRAM_GET_MTU`. Isn't this effectively the same?

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 10 above)
Comment (mattcaswell) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:125]: Ah, right so this is really an SSL function. I find the BIO_* prefix used here as confusing.

File: ddd/ddd-05-mem-nonblocking.DISCUSSION.md (hunk 9 above)
Comment (hlandau) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:103]: The BIO_dgram_mem will be passed to the SSL object, and we can certainly expect the SSL object to make use of the MTU set on the dgram_mem. Since the application is responsible for setting the MTU value on the dgram_mem, it doesn't really make sense for it to query it from the same dgram_mem.

This is more about minimising surprises on the application side.

Comment (mattcaswell) [reply ddd/ddd-05-mem-nonblocking.DISCUSSION.md:103]: Somehow I'm still missing something here...

File: ddd/ddd-02-conn-nonblocking.DISCUSSION.md (hunk 4 above)
Comment (t8m) [reply ddd/ddd-02-conn-nonblocking.DISCUSSION.md:122]: > Whether it warrants deprecation, I'm not sure.

Unless the name is confusing by itself (i.e., even if used for DTLSv1) I would not deprecate it. I'd just keep both names.
